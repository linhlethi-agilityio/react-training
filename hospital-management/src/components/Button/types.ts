import { ButtonHTMLAttributes, ReactNode } from 'react';
import { Variant } from '@data-types/variant';

export interface IButtonProps extends ButtonHTMLAttributes<HTMLButtonElement> {
  variant: Variant;
  children: ReactNode;
}
