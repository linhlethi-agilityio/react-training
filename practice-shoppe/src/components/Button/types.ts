import { BaseVariant } from '@type/variant';
import { ButtonHTMLAttributes, ReactNode } from 'react';

export type Variant = BaseVariant | 'outline';

export interface IButtonProps extends ButtonHTMLAttributes<HTMLButtonElement> {
  variant: Variant;
  children: ReactNode;
}
