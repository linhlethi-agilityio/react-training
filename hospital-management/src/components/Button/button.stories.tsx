import { ComponentMeta, ComponentStory } from '@storybook/react';
import { StyledButton } from '@components/Button';

export default {
  title: 'Button',
  component: StyledButton,
  argTypes: { onClick: { action: 'clicked' } },
  parameters: {
    backgrounds: {
      default: 'white',
      values: [
        { name: 'blue', value: '#283779' },
        { name: 'white', value: '#FFFFFF' },
      ],
    },
  },
} as ComponentMeta<typeof StyledButton>;

const Template: ComponentStory<typeof StyledButton> = (args) => <StyledButton {...args} />;

export const PrimaryButton = Template.bind({});
const primaryButton = {
  variant: 'primary',
  children: 'Button',
};
PrimaryButton.args = primaryButton;

export const SecondaryButton = Template.bind({});
const secondaryButton = {
  variant: 'secondary',
  children: 'Button',
};
SecondaryButton.args = secondaryButton;

export const RoundHeadButton = Template.bind({});
const roundHeadButton = {
  variant: 'roundHead',
  children: 'Button',
};
RoundHeadButton.parameters = {
  backgrounds: { default: 'blue' },
};
RoundHeadButton.args = roundHeadButton;
