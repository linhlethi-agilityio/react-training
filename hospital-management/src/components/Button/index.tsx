import { memo } from 'react';
import styled, { css } from 'styled-components';
import { IButtonProps } from './types';

const StyledButton = memo(
  styled.button<IButtonProps>`
    border-radius: 10px;
    padding: 7px 32px;
    font-size: var(--text-xxx-small);
    line-height: 18px;
    border: 1px solid var(--color-primary);
    display: inline-block;
    cursor: pointer;

    ${(props) =>
      props.variant === 'secondary' &&
      css`
        background-color: var(--color-primary);
        font-weight: var(--font-weight-bold);
        color: #ffffff;

        :hover {
          background-color: #376ba5;
        }
      `}

    ${(props) =>
      props.variant === 'primary' &&
      css`
        background-color: #ffffff;
        color: var(--color-primary);
        font-weight: var(--font-weight-normal);

        :hover {
          background-color: #e6e6e6;
        }
      `}

  ${(props) =>
      props.variant === 'roundHead' &&
      css`
        font-family: var(--font-secondary);
        background-color: #ffffff;
        color: #283779;
        border: none;
        padding: 14px 39px;
        font-size: var(--text-small);
        line-height: 19px;
        border-radius: 20px;
        font-weight: var(--font-weight-bold);

        :hover {
          background-color: #e6e6e6;
        }
      `}
  `,
  (prev, next) =>
    prev.variant === next.variant &&
    prev.disabled === next.disabled &&
    prev.onClick === next.onClick,
);

export { StyledButton };
