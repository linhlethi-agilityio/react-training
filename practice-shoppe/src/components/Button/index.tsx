import { memo } from 'react';
import styled, { css } from 'styled-components';
import { IButtonProps } from './types';

const StyledButton = memo(
  styled.button<IButtonProps>`
    font-weight: var(--font-weight-bold);
    line-height: 21px;
    cursor: pointer;
    text-transform: uppercase;

    ${(props) =>
      props.variant === 'secondary' &&
      css`
        background-color: #ffffff;
        color: var(--color-primary);
        padding: 16px 136px;
        border: 1px solid #000000;
        border-radius: 4px;
        font-size: var(--text-small);

        :hover {
          background-color: #e6e6e6;
        }
      `}

    ${(props) =>
      props.variant === 'primary' &&
      css`
        background-color: var(--color-primary);
        color: #ffffff;
        padding: 16px 125px;
        border: 1px solid #000000;
        border-radius: 4px;
        font-size: var(--text-small);

        :hover {
          background-color: #333333;
        }
      `}
    }

    ${(props) =>
      props.variant === 'outline' &&
      css`
        background-color: #efefef;
        width: 50px;
        height: 53px;
        border: none;
        color: var(--color-tertiary);
        font-size: var(--text-small);
      `}

    :disabled {
      background-color: #efefef;
      cursor: no-drop;
    }
  `,
);

export { StyledButton };
