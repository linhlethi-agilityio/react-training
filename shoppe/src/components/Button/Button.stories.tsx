import { ComponentMeta, ComponentStory } from '@storybook/react';
import { StyledButton } from '@components/Button';

export default {
  title: 'Button',
  component: StyledButton,
  argTypes: { onClick: { action: 'clicked' } },
} as ComponentMeta<typeof StyledButton>;

const Template: ComponentStory<typeof StyledButton> = (args) => <StyledButton {...args} />;

export const PrimaryButton = Template.bind({});
const primaryButton = {
  variant: 'primary',
  children: 'Button',
};
PrimaryButton.args = primaryButton;

export const SecondaryButton = Template.bind({});
const secondaryButton = {
  variant: 'secondary',
  children: 'Button',
};
SecondaryButton.args = secondaryButton;

export const DisabledButton = Template.bind({});
const disabledButton = {
  variant: 'secondary',
  children: 'Button',
  disabled: true,
};
DisabledButton.args = disabledButton;
