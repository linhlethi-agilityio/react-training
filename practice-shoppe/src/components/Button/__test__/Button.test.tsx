import { StyledButton } from '@components/Button';
import { fireEvent, render } from '@testing-library/react';

describe('testing Button component', () => {
  const onClick = jest.fn();

  it('testing secondary variant props', () => {
    const { asFragment } = render(<StyledButton variant="secondary">Button</StyledButton>);

    expect(asFragment()).toMatchSnapshot();
  });

  it('testing primary variant props', () => {
    const { asFragment } = render(<StyledButton variant="primary">Button</StyledButton>);

    expect(asFragment()).toMatchSnapshot();
  });

  it('testing outline variant props', () => {
    const { asFragment } = render(<StyledButton variant="outline">Button</StyledButton>);

    expect(asFragment()).toMatchSnapshot();
  });

  test('should render button correctly', () => {
    const { getByRole } = render(<StyledButton variant="outline">Button</StyledButton>);
    const button = getByRole('button', { name: /button/i });
    expect(button).toBeTruthy();
  });

  test('should call onClick one time when click button', () => {
    const { getByRole } = render(
      <StyledButton variant="outline" onClick={onClick}>
        Button
      </StyledButton>,
    );
    const button = getByRole('button', { name: /button/i });
    fireEvent.click(button);
    expect(onClick).toBeCalledTimes(1);
  });
});
