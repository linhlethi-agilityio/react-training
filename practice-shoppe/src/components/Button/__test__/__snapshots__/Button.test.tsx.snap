// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing Button component testing outline variant props 1`] = `
<DocumentFragment>
  <button
    class="sc-beqWaB fiuVov"
  >
    Button
  </button>
</DocumentFragment>
`;

exports[`testing Button component testing primary variant props 1`] = `
<DocumentFragment>
  <button
    class="sc-beqWaB jDRNRg"
  >
    Button
  </button>
</DocumentFragment>
`;

exports[`testing Button component testing secondary variant props 1`] = `
<DocumentFragment>
  <button
    class="sc-beqWaB IlpDp"
  >
    Button
  </button>
</DocumentFragment>
`;
